### ISSUE TITLE
full demo example?

### ISSUE THREAD

**Issue by vsoch** [smartfile/django-transfer#9]
hey @smartfile! I'm having a lot of trouble (as in, many weeks of angst) trying to get a chunked upload to work for django. I've tried all the django chunked upload modules (the web UI with jquery works ok, but I couldn't translate this to a command line version) and now I'm trying to get the nginx module working. I first was doing it without django-transfer and getting 502 errors pretty consistently, and I found your module here and am hoping it can help. I can't figure out how to best configure it - now I'm getting 404s. Do you have even a simple working demo, anywhere? The links you referenced are for ruby. Thanks!

**Comment by btimby** [comment]
This is a very late reply, but chunked uploading via nginx with the referenced `'upload_module` won't work. The module is old and also only for POST (`multipart/form-data`).

I have a LUA module named `chunkedup` that will handle `Transfer-Encoding: chunked`, `PUT` requests by writing them to disk and then passing a `POST` to your backend with the file details (including the temp file path). It is compatible with this django app. The project has a working example that you can `POST` and `PUT` to, it echos the `POST` request that _would_ be sent to your backend. _THAT_ `POST` request is what this app will convert into `request.FILES[]` in django, so you don't have to treat normal posts and accelerated ones any differently.

https://github.com/btimby/nginx-lua-chunkedup

Basically, this django app is just an adapter that converts the `POST` with file details provided by the upload module into what looks like a regular POST upload directly to django. It just means you don't have to do anything special to work with the upload module you choose.

You still have to configure the upload module and that is beyond the scope of this project.

**Comment by vsoch** [comment]
hey @btimby - no worries about the late reply, it's better than never!

I wound up getting the nginx large file upload (with the upload module) working for Singularity Registry Server (see Dockerfile [here](https://github.com/singularityhub/sregistry/blob/master/nginx/Dockerfile) and nginx.conf [here](https://github.com/singularityhub/sregistry/blob/master/nginx.conf#L35)) and your timing is uncanning because I'm circling back to try and support a multi part upload from the Singularity container client (it's a massive binary) and this particular upload module doesn't work because Singularity doesn't provide a Content Type header.  It looks like your module is [handling this](https://github.com/btimby/nginx-lua-chunkedup/blob/master/lua/nginx_lua_chunkedup.lua#L15) however. I'd like to try it out, but I have a few quick questions:

 - how would it handle authorization? Other than writing the file to disk and then handing off to a view (which would need to check if the upload was allowed and then proceed or not) in practice how have you protected these endpoints from DDOS attacks or similar?
 - you are saying I can implement nginx-lua-chunkedup on my own alongside my Django server, without needing django-transfer?

**Comment by btimby** [comment]
Firstly, authorization.

By default, authorization is handled AFTER the upload, during the POST to the backend. The backend is not involved until the upload has completed. This can be bad for nginx handling uploads that are eventually refused. The module cleans up the temp files after a non-200 error code.

However, because the module is LUA, you could add a snippet of code to perform authorization beforehand. Just make a subrequest to the backend and pass along the session cookie or `Authorization` header. Openresty has different phases of the request where LUA can be executed. You would have to refer to the documentation for specifics.

You could also use another module to authorize the request such as:
https://nginx.org/en/docs/http/ngx_http_auth_request_module.html#auth_request

The options are endless, and the solution will depend on your specific requirements.

Second, `django-transfer`  depends on a module like `upload_module` or `nginx-lua-upload-module` or `nginx-lua-chunkedup` to provide the POST request with a temp file (rather than file in body). Not the other way around. These upload modules can certainly be used without `django-transfer`. The benefit of `django-transfer` is that you don't need to handle the POST any differently than you would in the absence of the upload module. Put another way, the Django file upload example will work unchanged with `django-transfer` even when using an upload module.

This is just a small detail. You can certainly just handle the post and read the `content-type`, `file-path` etc. fields that the upload module provides and handle the upload directly.

**Comment by btimby** [comment]
I will try to explain another way... in code...

Without `django-transfer`:

```python
def my_upload_handler(request):
  # Is this a standard upload?
  if request.FILES:
    # yes, save the file...
    f = request.FILES['file']
    f.move('somewhere')
  else:
    # No, this upload came from our upload module.
    shutil.move(request.POST['[filename]'')
```

With `django-transfer`:

```python
def my_upload_handler(request):
    # All uploads are "normal" uploads...
    f = request.FILES['file']
    f.move('somewhere')
```

**Comment by btimby** [comment]
The reason I wrote this thing long, long ago was because during development, we did not have nginx in front of the application. In that case, the upload hit our django view directly. But when deployed, nginx handled the file upload and sent a different type of POST to the django view. Rather than writing logic to handle both cases, I created `django-transfer` to make both cases the same. Then the application code could be simpler and handle both cases.

Now with docker, this is not so important, you can have nginx on your development stack, but another reason this could be useful is if you have a django app already that handles uploads and you want to start offloading those uploads to nginx. With `django-transfer` you can do this with no changes to your views, just install this app and set up nginx.

It is just a compatibility layer.

**Comment by btimby** [comment]
Honestly, if your problem is a missing `Content-Type` header, you can use a module like `nginx-headers-more` to append that header outside the upload module.

**Comment by vsoch** [comment]
This is so great! You've given me multiple things to try, and I'm actually excited about that :) I'm going to start with trying out headers-more-nginx-module (since that is the least number of changes) and then I'll build a nginx container with the lua extension, and then play around with how that works. I'll report back with questions or updates!

**Comment by btimby** [comment]
Least number of changes is the best number of changes! Good luck.

**Comment by btimby** [comment]
Well, the upload module wants to see `Content-Type: multipart/form-data`, that is what it parses. It reads the form and saves any included files to temporary files.

It _then_ creates another POST request, still `Content-Type: multipart/form-data` to your backend `/api/uploads/complete/`. The only difference being that it modified the file type fields to be regular fields containing size, path, content-type of the uploaded file etc.

Is your problem that Sylabs (I am not familiar with this at all) does not set the outbound `Content-Type` header going back to the client _after_ the download? Is Sylabs at `/api/uploads/complete`?

**Comment by vsoch** [comment]
Changing the content type indeed fixed the error - the Post goes through (and I'll need to work on that now to figure out what exactly is going through). 

To answer your question, Sylabs implements [this protocol](https://docs.aws.amazon.com/AmazonS3/latest/dev/mpuoverview.html) intended for multipart uploads to S3. I'm trying to figure out how to do something similar for Singularity Registry server, so 1) it works with `singularity push` and 2) it also handles large files. My current solution works well with the singularity push library client, however it's limited on memory - on my machine I can push maybe up to between 5-8GB images (depending on what else I have running) so you can imagine to serve as a registry for a cluster this wouldn't scale well. For Singularity Hub that I developed and still maintain, it's easy to handle this because the builder instance (a Google Cloud instance) is the one that uploads directly to storage, and a download for an image pings Singularity Hub and then returns a signed URL for storage. I want to provide a scalable solution for Singularity Registry Server too (and also develop it for Django OCI, the Opencontainers distribution specification I mentioned for Django) within the constraints of what many centers want, which is a fully internal (no cloud services) solution.

**Comment by vsoch** [comment]
For this particular change, the POST reports EOF
```
$ singularity --debug -vvvv push  -U big.sif library://vsoch/test/busybox:latest
DEBUG   [U=1000,P=17288]   persistentPreRun()            Singularity version: 3.5.2+348-g72b09f23b-dirty
DEBUG   [U=1000,P=17288]   handleConfDir()               /home/vanessa/.singularity already exists. Not creating.
DEBUG   [U=1000,P=17288]   handleRemoteConf()            Ensuring file permission of 0600 on /home/vanessa/.singularity/remote.yaml
WARNING [U=1000,P=17288]   LibraryPush()                 Skipping container verifying
FATAL   [U=1000,P=17288]   func50()                      Unable to push image to library: error making request to server:
	Post http://127.0.0.1/v2/imagefile/15/_multipart: EOF
```

**Comment by vsoch** [comment]
Another idea I had this morning was to try adding a [ceph](https://singularityhub.github.io/sregistry-cli/client-swift) storage backend (which I believe uses the same protocol as S3) as another container deployed with the registry, and then those APIs would handle the upload (push) and download (pull).

**Comment by btimby** [comment]
Seems like your post is going through but the client expects some sort of body returned from the request.

For another S3 workalike, look at:

https://docs.min.io/docs/minio-docker-quickstart-guide.html

**Comment by btimby** [comment]
Also:

https://github.com/adobe/S3Mock

**Comment by btimby** [comment]
I am going to close this issue as I don't think it is relevant to `django-transfer` anymore :-).

**Comment by vsoch** [comment]
Understandable. Thanks for your (much extra) help!

**Comment by vsoch** [comment]
Hey @btimby thanks again for your help! I wound up implementing a minio backend and the functions plug in beautifully https://github.com/singularityhub/sregistry/pull/297. I tested a ~4.5GB container and it doesn't seem to have issues, but I'm really happy with this approach because 1) it plugs into the Singularity client fairly seamlessly, 2) we can use signed GET and PUT urls instead of doing validation on the server after the file has been uploaded, 3) there is better potential to scale the storage since the user can deploy their own Minio (outside the docker-compose) and just update the credentials and hostname (parameters in sregistry). Based on reading [here](https://github.com/minio/minio-js/issues/772) I might need to do a little more work to support the multipart upload (the example there is unfortunately for javascript) I will need to dig into the Sylabs source code to see how it's handling the multiupload part, and how I can interface with minio, I found [an example](https://github.com/T-Tmnr/aws_s3_multipartupload_with_presigned_url/blob/927685c378c4e929ce606dbec735304a3e49a0bd/aws-multipart-upload.py) that I'm hoping will be at least a start. But probably not today, going to work on other things.
